vm/systemSmartContracts/delegation.go

```diff
@@ -2599,6 +2615,128 @@ func (d *delegation) getMetaData(args *vmcommon.ContractCallInput) vmcommon.Retu
+	if !d.flagLiquidStaking.IsSet() {
+		d.eei.AddReturnMessage(args.Function + " is an unknown function")
+		return vmcommon.UserError
+	}
+	if !bytes.Equal(args.CallerAddr, vm.LiquidStakingSCAddress) {
+		d.eei.AddReturnMessage("only liquid staking sc can call this function")
+		return vmcommon.UserError
+	}
+	if args.CallValue.Cmp(zero) != 0 {
+		d.eei.AddReturnMessage("call value must be 0")
+		return vmcommon.UserError
+	}
+	if len(args.Arguments) < 2 {
+		d.eei.AddReturnMessage("not enough arguments")
+		return vmcommon.UserError
+	}
+	err := d.eei.UseGas(d.gasCost.MetaChainSystemSCsCost.DelegationOps)
+	if err != nil {
+		d.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+	address := args.Arguments[0]
+	value := big.NewInt(0).SetBytes(args.Arguments[1])
+	if value.Cmp(zero) <= 0 {
+		d.eei.AddReturnMessage("invalid argument for value as bigInt")
+		return vmcommon.UserError
+	}
+	if len(address) != len(d.validatorSCAddr) {
+		d.eei.AddReturnMessage("invalid address as input")
+		return vmcommon.UserError
+	}
+
+	return vmcommon.Ok
+}
+
+func (d *delegation) claimDelegatedPosition(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	returnCode := d.basicCheckForLiquidStaking(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+
+	address := args.Arguments[0]
+	value := big.NewInt(0).SetBytes(args.Arguments[1])
+
+	isNew, delegator, err := d.getOrCreateDelegatorData(address)
+	if err != nil {
+		d.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+	if isNew {
+		d.eei.AddReturnMessage("caller is not a delegator")
+		return vmcommon.UserError
+	}
+
+	activeFund, err := d.getFund(delegator.ActiveFund)
+	if err != nil {
+		d.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+	if value.Cmp(activeFund.Value) > 0 {
+		d.eei.AddReturnMessage("not enough funds to claim position")
+		return vmcommon.UserError
+	}
+
+	activeFund.Value.Sub(activeFund.Value, value)
+	err = d.saveFund(delegator.ActiveFund, activeFund)
+	if err != nil {
+		d.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	err = d.deleteDelegatorIfNeeded(args.CallerAddr, delegator)
+	if err != nil {
+		d.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	d.eei.Finish(big.NewInt(int64(delegator.RewardsCheckpoint)).Bytes())
+	return vmcommon.Ok
+}
+
+func (d *delegation) claimRewardsViaLiquidStaking(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	returnCode := d.basicCheckForLiquidStaking(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+
+	if len(args.Arguments) != 3 {
+		d.eei.AddReturnMessage("not enough arguments")
+		return vmcommon.UserError
+	}
+
+	address := args.Arguments[0]
+	value := big.NewInt(0).SetBytes(args.Arguments[1])
+	checkPoint := uint32(big.NewInt(0).SetBytes(args.Arguments[2]).Uint64())
+
+	totalRewards, err := d.computeRewards(checkPoint, false, value)
```

Comment:
so the claim operation via liquid staking will also take all unclaimed rewards (if existing) before creating the liquid position, right?

Reply:
yes.

Reply:
good